File: Sorting algorithms, heap, bs/F.java
import java.util.Scanner;

public class NearBinSearch {

    static long[] a;

    static long bs(int l, int r, long x){
        int m = 0;
        while (r-l>1){
            m = (l+r)/2;
            if (a[m]<=x){
                l = m;
            } else {
                r = m;
            }
        }
        if (l<0) return a[r];
        else if(r>=a.length) return a[l];
        else if (Math.abs(x-a[r])<Math.abs(x- a[l])){
            return a[r];
        } else return a[l];
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt(), k = scanner.nextInt();
        a = new long[n];
        for (int i = 0; i< n; i++){
            a[i] = scanner.nextLong();
        }
        for (int i = 0; i < k; i++){
            long z = bs(-1, n, scanner.nextInt());
            System.out.println(z);
        }
    }
}

File: DFS, MST ITMO aignments/F.kt
import java.util.Scanner

val scanner = Scanner(System.`in`)

val n = scanner.nextInt()

val m = scanner.nextInt()

val graph = Array(n) { Vertex() }

val list = mutableListOf<Int>()

class Vertex(
        var used: Int = 0,
        var color: Int = 0,
        val edges: MutableList<Int> = mutableListOf(),
        val edgesT: MutableList<Int> = mutableListOf()
)

fun main() {
    var color = 0
    val r = Array(10_000) { mutableSetOf<Int>() }
    for (i in 1..m) {
        val u = scanner.nextInt() - 1
        val v = scanner.nextInt() - 1
        graph[u].edges.add(v)
        graph[v].edgesT.add(u)
    }
    for (i in 0 until n) if (graph[i].used != 1) dfsG(i)
    for (i in list.reversed()) if (graph[i].used != 2) dfsTG(i, color++)
    for (i in 0 until n) for (j in graph[i].edges) if (graph[i].color != graph[j].color)
            r[graph[i].color].add(graph[j].color)
    var ans = 0
    r.forEach { ans += it.size }
    println(ans)
}

fun dfsG(v: Int) {
    graph[v].used = 1
    for (i in graph[v].edges) if (graph[i].used != 1) dfsG(i)
    list.add(v)
}

fun dfsTG(v: Int, color: Int) {
    graph[v].used = 2
    graph[v].color = color
    for (i in graph[v].edgesT) if (graph[i].used != 2) dfsTG(i, color)
}
